Fix action handler typing

altano opened the pull request:
Action handlers can currently have invalid states and actions. There are unit tests to prevent this but unfortunately the `@ts-expect-error` was catching unrelated issues, and these parameters aren't currently checked. I've gone ahead and made the fixes required to type check the state and action here correctly.

The only user-facing change here is that I had to change `ErrorBrand`'s implementation and this will affect user-facing errors (but I hope you will agree they're more readable now).

rickwebiii commented on src/StateMachine.ts:
```diff
@@ -31,7 +31,7 @@ export type Action<Name extends ActionNameType, Payload> = Readonly<Payload> & {
  * issue is rather than something puzzling like could not assign to never.
  */
-type ErrorBrand<T extends IndexType> = { [k in T]: void };
+type ErrorBrand<T> = T | never
```
`ErrorBrand` now does nothing. `T | never` is simply `T`. The brand is designed to make string literal types unassignable to errors, which this no longer does.

The branding is needed to prevent some sneaky edge cases. E.g. a user names a state, action or whatever to be an error message: 'Foo has already been declared'.

rickwebiii commented:
Hmm... the test cases I looked at appear to fail as expected without this change. The only difference I can see is that with your change, the input types don't widen (e.g. `'b'` doesn't become `string`). The error messages are clearer with these changes, but unfortunately the branding is needed to prevent users from doing sneaky things like making a state with the name of an error message, which causes an extends check to pass and breaks the error checking.

Do you have a particular test case I can look at to see the impact of these changes?

rickwebiii commented on src/StateMachine.ts:
```diff
@@ -268,9 +270,9 @@ const action = <StateMap, Transitions, ActionMap>(): ActionFunc<StateMap, Transi
-    action: AN,
+  return <S extends StateType, AN extends ActionNameType, NS extends MapValues<StateMap>>(
+    state: AssertStateInMap<StateMap, S>,
+    action: AssertActionIsDefined<AN, MapKeys<ActionMap>>,
```
Ah ok, I think this will give a better error message and I think this might have slipped through the cracks.

rickwebiii commented on test/StateMachine.ts:
```diff
@@ -154,42 +154,55 @@ describe('compile-time checking', () => {
-            // @ts-expect-error
-            .state<'b'>('b');
+            .state<'b'>(
+                // @ts-expect-error
```
👍 better scoping the assertion

rickwebiii commented on test/StateMachine.ts:
```diff
@@ -233,13 +253,16 @@ describe('compile-time checking', () => {
+                'a2',
+                () => {
+                    return {
+                        stateName: 'b' as const
```
Ok, I see now, this is indeed an oversight. This case isn't caught without your change and the reason for the failure is the missing const assertion, not the undefined action.

rickwebiii commented:
Ok, I see now. `should fail when declaring handler to non-action` wasn't actually catching a genuine error because it was missing an `as const`. Overall, I think if you revert the branding change, everything else can be taken as-is.

Thanks!

altano commented:
I thought my branding change handled the case where you specified an error message string to circumvent the type checking but it wasn't actually working. Whoops.

I've
- Replaced my `ErrorBrand` change with one that adds unit tests to catch my mistake so no one else makes it in the future
- Changed the code to cast the state and action to get rid of the error about `ErrorBrand` being in the type. This isn't the nicest code but you also had to do it elsewhere in the file already, and it's something we can clean up if throw types ever land in TypeScript.

Let me know if this looks good now, and if not, what else I can do to fix it up to your liking. Thanks again!

rickwebiii reviewed (approved):
LGTM. I'll go ahead and merge this.
